libgnucash/engine/Account.cpp

```diff
@@ -2691,6 +2691,45 @@ DxaccAccountSetCurrency (Account * acc, gnc_commodity * currency)
 /********************************************************************\
 \********************************************************************/
 
+static gpointer
+account_foreach_descendant (const Account *acc, AccountCb2 thunk,
+                            gpointer user_data, gboolean sort, gboolean can_abort)
+{
+    const AccountPrivate *priv;
+    gpointer result = nullptr;
+    GList *children;
+
+    g_return_val_if_fail (GNC_IS_ACCOUNT(acc), nullptr);
+    g_return_val_if_fail (thunk, nullptr);
+
+    priv = GET_PRIVATE(acc);
+    if (sort)
+    {
+        children = g_list_copy (priv->children);
+        children = g_list_sort (children, (GCompareFunc)xaccAccountOrder);
+    }
+    else
+        children = priv->children;
+
+    for (auto node = children; node; node = node->next)
+    {
+        auto child = static_cast<Account*>(node->data);
+        result = thunk (child, user_data);
+        if (can_abort && result) break;
+
+        result = account_foreach_descendant (child, thunk, user_data, sort, can_abort);
+        if (can_abort && result) break;
```

Comment:
This is going one step too far for code reuse. `for_each` means that a thunk's side effects are applied to every node in the collection. Make a separate `gnc_account_find_first` that returns the desired value and call the function argument `predicate`.

Reply:
Will continue later. I presume `gnc_account_find_first` will call `account_foreach_descendant` with `can_abort` being `TRUE`.

Reply:
It would be better to just separate them. `foreach` whatever should always touch every `each`, no aborting. There's also no need for the sort logic in `find_first` so the duped code is down to the precondition check, 1thunk,(child)` and the recursion loop. One flavor returns, the other maybe sorts. Not that much duplication.

Reply:
Ok in this case `gnc_account_foreach_descendant_until` doesn't need to change that much, just needs using C++ syntax. Pushed modifying changes and will merge commits to appropriate counterparts if satisfactory.